crowdsourcehinter/static/js/src/crowdsourcehinter.js

```diff
@@ -0,0 +1,305 @@
+function CrowdsourceHinter(runtime, element, data){
+
+    var executeHinter = true;
+    var isShowingHintFeedback = false;
+    var voted = false;
+    var correctSubmission = false;
+    
+    $(".crowdsourcehinter_block", element).hide();
+
+    if(!executeHinter){
+        return;
+    }
+
+    /**
+     * Set executeHinter to false, disabling the hinter xblock. Triggered by switching units
+     * in edX course.
+     * This is a workaround for when a student switches to/from a unit and causes multiple
+     * instances of the hinter to be running.
+     */
+    function stopScript(){
+        executeHinter = false;
+    }
+    Logger.listen('seq_next', null, stopScript);
+    Logger.listen('seq_prev', null, stopScript);
+    Logger.listen('seq_goto', null, stopScript);
+
+    /**
+     * Get a hint to show to the student after incorrectly answering a question.
+     * @param data is data generated by the problem_graded event
+     */
+    function get_hint(data){
+        $(".crowdsourcehinter_block", element).show();
+        $.ajax({
+            type: "POST",
+            url: runtime.handlerUrl(element, 'get_hint'),
+            data: JSON.stringify({"submittedanswer": unescape(data[0])}),
+            success: showHint
+        });
+    }
+
+    /**
+     * Start student hint feedback. This function is called after the student answers
+     * the question correctly.
+     */
+    function start_feedback(){
+        $('.csh_correct', element).show();
+        $(".csh_hint_reveal", element).hide();
+        //send empty data for ajax call because not having a data field causes error
+        $.ajax({
+            type: "POST",
+            url: runtime.handlerUrl(element, 'get_feedback'),
+            data: JSON.stringify({}),
+            success: showStudentContribution
+        });
+    }
+
+    /**
+     * Check whether student answered the question correctly and call the appropriate
+     * function afterwards. Current method for determining correctness if very brittle.
+     * @param event_type, element are both unused but automatically passed
+     * @param data is generated by problem_graded event, contains status and data of the problem block
+     */
+    function onStudentSubmission(){ return function(event_type, data, element){
+        //search method of correctness of problem is brittle due to checking for a class within
+        //the problem block.
+        if (data[1].search(/class="correct/) === -1){
+            get_hint(data);
+        } else { //if the correct answer is submitted
+            start_feedback();
+        }
+    }}
+    Logger.listen('problem_graded', data.hinting_element, onStudentSubmission());
+
+    /**
+     * Modify csh_Hints attributes to show hint to the student.
+     */
+    function showHint(result){
+        $('.csh_Hints', element).attr('student_answer', result.StudentAnswer);
+        $('.csh_Hints', element).attr('hint_received', result.Hints);
```

Comment:
What is hints? I'm wondering why this is plural. It feels singular from the code. Regardless, it'd be helpful if the comment showed an actual sample bit of JSON: 

    {
         'StudentAnswer' : "7",
         'Hints' : "You had an off-by-one error"
    }

(note: I have no idea what the actual JSON is, so don't cut-and-paste the above.